core/src/sumeragi/fault.rs

```diff
@@ -294,8 +262,133 @@ impl<F: FaultInjection> SumeragiWithFault<F> {
+        *wsv_for_public_use_guard = WorldStateView::clone(&state_machine_guard.wsv);
+    }
+
+    for event in Vec::<Event>::from(&block) {
+        trace!(?event);
+        drop(sumeragi.events_sender.send(event));
+    }
+
+    state_machine_guard.latest_block_height = block.header().height;
+    state_machine_guard.latest_block_hash = block.hash();
+
+    // Push new block height information to block_sync
+    *sumeragi
+        .latest_block_hash_for_use_by_block_sync
+        .lock()
+        .expect("lock on latest_block_hash_for_use_by_block_sync") =
+        state_machine_guard.latest_block_hash;
+
+    let previous_role = state_machine_guard.current_topology.role(&sumeragi.peer_id);
+    state_machine_guard
+        .current_topology
+        .refresh_at_new_block(block_hash);
+    info!(
+        prev_peer_role = ?previous_role,
+        new_peer_role = ?state_machine_guard.current_topology.role(&sumeragi.peer_id),
+        new_block_height = %state_machine_guard.latest_block_height,
+        %block_hash,
+        "Committing block"
+    );
+    sumeragi.kura.blocking_store_block(block);
+    SumeragiWithFault::<F>::update_network_topology(
+        &mut state_machine_guard.current_topology,
+        &state_machine_guard.wsv,
+    );
+
+    // Transaction Cache
+    {
+        let mut transaction_cache = &mut state_machine_guard.transaction_cache;
+        // We prune transactions now because now is the only time they may have been
+        // invalidated by a new block being applied to the world state view.
+        //
+        // We also check for if the transaction has expired because we might as well.
+        let mut read_index = 0;
+        let mut write_index = 0;
+        while read_index < transaction_cache.len() {
+            if transaction_cache[read_index].is_some() {
+                let tx = transaction_cache[read_index].take().unwrap();
+                if tx.is_in_blockchain(&state_machine_guard.wsv)
+                    || tx.is_expired(sumeragi.queue.ttl)
+                {
+                    read_index += 1;
+                    continue;
+                }
+                transaction_cache[write_index] = Some(tx);
+                read_index += 1;
+                write_index += 1;
+                continue;
+            }
+            read_index += 1;
+        }
+        transaction_cache.truncate(write_index);
+    }
+}
+
+fn request_view_change<F>(
+    sumeragi: &SumeragiWithFault<F>,
+    state_machine_guard: &mut SumeragiStateMachineData,
+    view_change_proof_chain: &mut Vec<Proof>,
+    current_view_change_index: u64,
+) where
+    F: FaultInjection,
+{
+    let mut suspect_proof = Proof {
+        latest_block_hash: state_machine_guard.latest_block_hash,
+        view_change_index: current_view_change_index,
+        signatures: Vec::new(),
+    };
+    suspect_proof
+        .sign(sumeragi.key_pair.clone())
+        .expect("must be able to perform signing");
```

Comment:
```suggestion
    let suspect_proof = Proof {
        latest_block_hash: state_machine_guard.latest_block_hash,
        view_change_index: current_view_change_index,
        signatures: Vec::new(),
    }.sign(sumeragi.key_pair.clone())
     .expect("signing of suspect proof failed");
```

Reply:
Sorry. The API does not work that way. `sign` inserts signatures into an already existing proof. That is how the proof "votes" are accumulated across peers. And converting it to a functional style would make the borrow checker go off.